internal/domain/power-status.go

```diff
@@ -18,16 +19,707 @@
+// here.
+
+func updateVaultCreds() error {
+	tmpMap := make(map[string]*componentPowerInfo)
+
+	for k,v := range(hwStateMap) {
+		if ((v.BmcUsername == "") || (v.BmcPassword != "")) {
+			tmpMap[k] = v
+		}
+	}
+
+	err := getVaultCredsAll(tmpMap)
+	return err
+}
+
+func getVaultCredsAll(compMap map[string]*componentPowerInfo) error {
+	var un,pw string
+	var err error
+	fname := "getVaultCredsAll()"
+
+	if (!vaultEnabled || (ccStore == nil)) {
+		glogger.Warnf("%s: Vault is disabled.",fname)
+		return nil
+	}
+
+	//The credstore layer caches creds, so this should be fast.
+
+	for k,v := range(compMap) {
+		un,pw,err = (*ccStore).GetControllerCredentials(k)
+		if (err != nil) {
+			return fmt.Errorf("ERROR: Can't get BMC creds for '%s': %v",
+						k,err)
+		}
+		if ((un == "") || (pw == "")) {
+			glogger.Warnf("%s: Missing/empty creds for '%s'",fname,k)
+		}
+		v.BmcUsername = un
+		v.BmcPassword = pw
+	}
+
+	return nil
 }
+
+// Return the HTTP status code from a completed TRS task.  This is a little
+// funky... if a valid transaction yields no response, there will be no
+// Response structure.  This would basically be a 204.  But if there is an
+// error, you also get no Response struct, and thus no return code.  The
+// "Err" field should have something in it, but the messages don't give good
+// details.
+//
+// So for now, if there is no Response data and Err is nil, we will consider
+// that a 204.  If there is no Response and Err is populated, it will be a 500.
+
+func getStatusCode(tp *trsapi.HttpTask) int {
+	ecode := 600
+	if tp.Request.Response != nil {
+		ecode = int(tp.Request.Response.StatusCode)
+	} else {
+		if tp.Err != nil {
+			glogger.Tracef("getStatusCode, no response, err: '%v'", *tp.Err)
+			ecode = int(http.StatusInternalServerError)
+		} else {
+			ecode = int(http.StatusNoContent)
+		}
+	}
+	return ecode
+}
+
+// Fetch actual hardware status from hardware.
+
+func getHWStatesFromHW() error {
+	var url string
+	var powerState pcsmodel.PowerStateFilter
+
+	fname := "getHWStatesFromHW"
+	hashXName := http.CanonicalHeaderKey("XName")
+	hashCType := http.CanonicalHeaderKey("CType")
+	sourceTL := trsapi.HttpTask{Timeout: time.Duration(httpTimeout) * time.Second,}
+
+	//Get vault creds where needed
+
+	cerr := updateVaultCreds()
+	if (cerr != nil) {
+		return cerr
+	}
+
+	//TODO: maybe ask HBTD for current HB status of node elements, and if 
+	//they're heartbeating, set the status to ON and don't ask the HW.  This
+	//can have a pretty large window of error however.
+
+	//Use TRS to get all HW states.  Create a map so the TRS task completion
+	//notifications can map back to an XName.
+
+	taskList := (*tloc).CreateTaskList(&sourceTL,len(hwStateMap))
+	activeTasks := 0
+
+	for k,v := range(hwStateMap) {	//key is component XName, val is HSM RF EP info
+		if (v.PSComp.ManagementState == pcsmodel.ManagementStateFilter_unavailable.String()) {
+			continue
+		}
+
+		ctype := xnametypes.GetHMSType(k)
+		switch (ctype) {
```

Comment:
Missing types:
"ComputeModule" 
"RouterModule" 
"CabinetPDUPowerConnector"

Reply:
These types don't have a BMC "parent" concept in the data obtained from HSM, correct?   I'll ask this in the slack channel.   I left lots of component types out since we can't query their power states from the hardware (IIRC).

Reply:
ComputeModules and RouterModules are actually the slots for the compute blades and Rosetta switches. They are controlled by the cC.

ComputeModule ${cC}/redfish/v1/Chassis/BladeX
RouterModule ${cC}/redfish/v1/Chassis/PerifX

The CabinetPDUPowerConnector has a reference to RTS in its HSM Inventory/ComponentEndpoint

Reply:
That being said, some of these components do not have 'PowerState' fields via redfish and are simply marked 'Populated' or 'Ready'(non-node) in HSM such as NodeEnclosures and BMCs.